### ISSUE TITLE
Observing high contention in yb::server::HybridClock::Now() 

### ISSUE THREAD

**Issue by mchang31415** [YugaByte/yugabyte-db#48]
Grabbed the following from http://{hostname}:9000/pprof/contention
In the same workload as https://github.com/YugaByte/yugabyte-db/issues/47, seeing high number of cpu cycles spent in HybridClock::Now()
Looks like there might be a software bottleneck that could be optimized.

```
-----------

3515904	323 @ 00007f57a6271e60 00007f57a984612d 00007f57abe24671 00007f57abe25df5 00007f57abdc8a3f 00007f57ac0e0df0 00007f57ab166dc7 00007f57a7bf1d67 00007f57a7bf41b9 00007f57a6284b46 00007f57a251d314 00007f57a225ebed 0000000000000000
    @     0x7f57a6271e5f  SubmitSpinLockProfileData (yb/util/spinlock_profiling.cc:238)
    @     0x7f57a984612c  base::SpinLock::Unlock() (yb/gutil/spinlock.h:120)
    @     0x7f57a984612c  yb::simple_spinlock::unlock() (yb/util/locks.h:65)
    @     0x7f57a984612c  std::lock_guard<yb::simple_spinlock>::~lock_guard() (/home/centos/.linuxbrew-yb-build/Cellar/gcc/5.3.0/include/c++/5.3.0/mutex:383)
    @     0x7f57a984612c  yb::server::HybridClock::Now() (yb/server/hybrid_clock.cc:218)
    @     0x7f57abe24670  operator() (yb/tablet/mvcc.cc:173)
    @     0x7f57abe24670  wait<yb::tablet::MvccManager::DoGetSafeTime(yb::HybridTime, yb::MonoTime, yb::HybridTime, std::unique_lock<std::mutex>*) const::<lambda()> > (/home/centos/.linuxbrew-yb-build/Cellar/gcc/5.3.0/include/c++/5.3.0/condition_variable:97)
    @     0x7f57abe24670  yb::tablet::MvccManager::DoGetSafeTime(yb::HybridTime, yb::MonoTime, yb::HybridTime, std::unique_lock<std::mutex>*) const (yb/tablet/mvcc.cc:190)
    @     0x7f57abe25df4  yb::tablet::MvccManager::SafeTime(yb::HybridTime, yb::MonoTime, yb::HybridTime) const (yb/tablet/mvcc.cc:161)
    @     0x7f57abdc8a3e  yb::tablet::Tablet::DoGetSafeTime(yb::StronglyTypedBool<yb::tablet::RequireLease_Tag>, yb::HybridTime, yb::MonoTime) const (yb/tablet/tablet.cc:1209)
    @     0x7f57ac0e0def  yb::tablet::AbstractTablet::SafeTime(yb::StronglyTypedBool<yb::tablet::RequireLease_Tag>, yb::HybridTime, yb::MonoTime) const (yb/tablet/abstract_tablet.h:72)
    @     0x7f57ac0e0def  yb::tserver::TabletServiceImpl::Read(yb::tserver::ReadRequestPB const*, yb::tserver::ReadResponsePB*, yb::rpc::RpcContext) (yb/tserver/tablet_service.cc:813)
    @     0x7f57ab166dc6  yb::tserver::TabletServerServiceIf::Handle(std::shared_ptr<yb::rpc::InboundCall>) (yb/tserver/tserver_service.service.cc:135)
    @     0x7f57a7bf1d66  yb::rpc::ServicePoolImpl::Handle(std::shared_ptr<yb::rpc::InboundCall>) (yb/rpc/service_pool.cc:191)
    @     0x7f57a7bf1d66  Run (yb/rpc/service_pool.cc:206)
    @     0x7f57a7bf1d66  Run (yb/rpc/tasks_pool.h:70)
    @     0x7f57a7bf41b8  Execute (yb/rpc/thread_pool.cc:98)
    @     0x7f57a6284b45  std::function<void ()>::operator()() const (/home/centos/.linuxbrew-yb-build/Cellar/gcc/5.3.0/include/c++/5.3.0/functional:2271)
    @     0x7f57a6284b45  yb::Thread::SuperviseThread(void*) (yb/util/thread.cc:602)

```

Full output here:
[contention.txt](https://github.com/YugaByte/yugabyte-db/files/1709367/contention.txt)

**Comment by kmuthukk** [comment]
@bmatican : assigning to you.

From the "contention.txt" file that was uploaded, it seems that two most prominent stacks in terms of spinlock content have HybridClock::Now(), are as follows:

Stack 1:
```
3515904	323 @ 00007f57a6271e60 00007f57a984612d 00007f57abe24671 00007f57abe25df5 00007f57abdc8a3f 00007f57ac0e0df0 00007f57ab166dc7 00007f57a7bf1d67 00007f57a7bf41b9 00007f57a6284b46 00007f57a251d314 00007f57a225ebed 0000000000000000
    @     0x7f57a6271e5f  SubmitSpinLockProfileData (yb/util/spinlock_profiling.cc:238)
    @     0x7f57a984612c  base::SpinLock::Unlock() (yb/gutil/spinlock.h:120)
    @     0x7f57a984612c  yb::simple_spinlock::unlock() (yb/util/locks.h:65)
    @     0x7f57a984612c  std::lock_guard<yb::simple_spinlock>::~lock_guard() (/home/centos/.linuxbrew-yb-build/Cellar/gcc/5.3.0/include/c++/5.3.0/mutex:383)
    @     0x7f57a984612c  yb::server::HybridClock::Now() (yb/server/hybrid_clock.cc:218)
    @     0x7f57abe24670  operator() (yb/tablet/mvcc.cc:173)
    @     0x7f57abe24670  wait<yb::tablet::MvccManager::DoGetSafeTime(yb::HybridTime, yb::MonoTime, yb::HybridTime, std::unique_lock<std::mutex>*) const::<lambda()> > (/home/centos/.linuxbrew-yb-build/Cellar/gcc/5.3.0/include/c++/5.3.0/condition_variable:97)
    @     0x7f57abe24670  yb::tablet::MvccManager::DoGetSafeTime(yb::HybridTime, yb::MonoTime, yb::HybridTime, std::unique_lock<std::mutex>*) const (yb/tablet/mvcc.cc:190)
    @     0x7f57abe25df4  yb::tablet::MvccManager::SafeTime(yb::HybridTime, yb::MonoTime, yb::HybridTime) const (yb/tablet/mvcc.cc:161)
    @     0x7f57abdc8a3e  yb::tablet::Tablet::DoGetSafeTime(yb::StronglyTypedBool<yb::tablet::RequireLease_Tag>, yb::HybridTime, yb::MonoTime) const (yb/tablet/tablet.cc:1209)
    @     0x7f57ac0e0def  yb::tablet::AbstractTablet::SafeTime(yb::StronglyTypedBool<yb::tablet::RequireLease_Tag>, yb::HybridTime, yb::MonoTime) const (yb/tablet/abstract_tablet.h:72)
    @     0x7f57ac0e0def  yb::tserver::TabletServiceImpl::Read(yb::tserver::ReadRequestPB const*, yb::tserver::ReadResponsePB*, yb::rpc::RpcContext) (yb/tserver/tablet_service.cc:813)
```

Stack 2:

-----------
```
1722112	100 @ 00007f57a6271e60 00007f57a984612d 00007f57ac0e0c44 00007f57ab166dc7 00007f57a7bf1d67 00007f57a7bf41b9 00007f57a6284b46 00007f57a251d314 00007f57a225ebed 0000000000000000
    @     0x7f57a6271e5f  SubmitSpinLockProfileData (yb/util/spinlock_profiling.cc:238)
    @     0x7f57a984612c  base::SpinLock::Unlock() (yb/gutil/spinlock.h:120)
    @     0x7f57a984612c  yb::simple_spinlock::unlock() (yb/util/locks.h:65)
    @     0x7f57a984612c  std::lock_guard<yb::simple_spinlock>::~lock_guard() (/home/centos/.linuxbrew-yb-build/Cellar/gcc/5.3.0/include/c++/5.3.0/mutex:383)
    @     0x7f57a984612c  yb::server::HybridClock::Now() (yb/server/hybrid_clock.cc:218)
    @     0x7f57ac0e0c43  yb::tserver::RpcOperationCompletionCallback<yb::tserver::ReadResponsePB>::OperationCompleted() (yb/tserver/service_util.h:146)
    @     0x7f57ac0e0c43  yb::tserver::TabletServiceImpl::Read(yb::tserver::ReadRequestPB const*, yb::tserver::ReadResponsePB*, yb::rpc::RpcContext) (yb/tserver/tablet_service.cc:878)
    @     0x7f57ab166dc6  yb::tserver::TabletServerServiceIf::Handle(std::shared_ptr<yb::rpc::InboundCall>) (yb/tserver/tserver_service.service.cc:135)
    @     0x7f57a7bf1d66  yb::rpc::ServicePoolImpl::Handle(std::shared_ptr<yb::rpc::InboundCall>) (yb/rpc/service_pool.cc:191)
    @     0x7f57a7bf1d66  Run (yb/rpc/service_pool.cc:206)
    @     0x7f57a7bf1d66  Run (yb/rpc/tasks_pool.h:70)
    @     0x7f57a7bf41b8  Execute (yb/rpc/thread_pool.cc:98)
    @     0x7f57a6284b45  std::function<void ()>::operator()() const (/home/centos/.linuxbrew-yb-build/Cellar/gcc/5.3.0/include/c++/5.3.0/functional:2271)
    @     0x7f57a6284b45  yb::Thread::SuperviseThread(void*) (yb/util/thread.cc:602)
```

**Comment by kmuthukk** [comment]
@mchang31415 : can you please share how many cores per yugabyte node in your setup?

**Comment by kmuthukk** [comment]
@bmatican - can we close out this issue with a link to the relevant commit?

**Comment by bmatican** [comment]
fixed in c62d02a8fdd2f0421dc5f91cf406ee2c805374d1
